crates/codegen/parser/generator/src/code_generator.rs

```diff
@@ -28,27 +28,31 @@ pub struct CodeGenerator {
     token_kinds: BTreeSet<&'static str>,
     trivia_kinds: BTreeSet<&'static str>,
 
-    top_level_scanner_names: BTreeSet<&'static str>,
-    scanner_functions: Vec<(&'static str, String)>, // (name of scanner, code)
-    scanner_contexts: Vec<ScannerContext>,
+    scanner_functions: BTreeMap<&'static str, String>, // (name of scanner, code)
+    scanner_contexts: BTreeMap<&'static str, ScannerContext>,
+    keyword_compound_scanners: BTreeMap<&'static str, String>, // (name of the KW scanner, code)
 
-    parser_functions: Vec<(&'static str, String)>, // (name of parser, code)
+    parser_functions: BTreeMap<&'static str, String>, // (name of parser, code)
 
     #[serde(skip)]
-    scanner_contexts_map: BTreeMap<&'static str, ScannerContext>,
+    top_level_scanner_names: BTreeSet<&'static str>,
     #[serde(skip)]
     all_scanners: BTreeMap<&'static str, ScannerDefinitionRef>,
     #[serde(skip)]
     current_context_name: &'static str,
 }
 
-#[derive(Serialize)]
+#[derive(Default, Serialize)]
 struct ScannerContext {
-    name: &'static str,
     #[serde(skip)]
     scanner_definitions: BTreeSet<&'static str>,
-    alpha_literal_scanner: String,
-    non_alpha_literal_scanner: String,
+    literal_scanner: String,
+    keyword_compound_scanners: BTreeMap<&'static str, String>,
+    keyword_trie_scanner: String,
+    #[serde(skip)]
+    keyword_scanner_defs: BTreeMap<&'static str, KeywordScannerDefinitionRef>,
+    /// Name of the scanners that can be promoted to a keyword
+    identifier_scanners: BTreeSet<&'static str>,
```

Comment:
If a name needs a comment to make up for the lack of sufficient naming, then something is wrong IMO. I believe that a long descriptive name is better than a short name that needs a comment.

Reply:
In a general context, I agree.

Here, however, the comment tries to enrichen the context, as keywords are already inherently "promotable" from the identifiers (no new information); this is just a reminder to the reader and hints how it's used later on in the template expansion. I'll remove it if you think it's redundant (the alternative of `identifier_scanner_names_promotable_to_keywords` seems a bit too extra and redundant, as per above).

Reply:
I've renamed from `identifier_scanners` to `identifier_scanner_names` in https://github.com/NomicFoundation/slang/pull/723/commits/581d611f291c64a7a9f7ee3133362ad105c78e67; this is more symmetrical with other `*_scanner_names` and carries some "new" context over from the comment above; hopefully that's a good middle ground!

Reply:
Done in https://github.com/NomicFoundation/slang/pull/723/commits/2336ad9538a04dcffaace368953272212bfa1a96